common/raw_hashtable.h

```diff
@@ -0,0 +1,1302 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_RAW_HASHTABLE_H_
+#define CARBON_COMMON_RAW_HASHTABLE_H_
+
+#include <algorithm>
+#include <cstddef>
+#include <cstring>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "common/check.h"
+#include "common/hashing.h"
+#include "common/raw_hashtable_metadata_group.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/Support/Compiler.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/MathExtras.h"
+
+// A namespace collecting a set of low-level utilities for building hashtable
+// data structures. These should only be used as implementation details of
+// higher-level data-structure APIs.
+//
+// These utilities support hashtables following a *specific* API design pattern,
+// and using Small-Size Optimization, or "SSO", when desired. We expect there to
+// be three layers to any hashtable design:
+//
+// - A *view* type: a read-only view of the hashtable contents. This type should
+//   be a value type and is expected to be passed by-value in APIs. However, it
+//   will have `const`-reference semantics, much like a `std::string_view`. Note
+//   that the *entries* will continue to be mutable, it is only the *table* that
+//   is read-only.
+//
+// - A *base* type: a base class type of the actual hashtable, which allows
+//   almost all mutable operations but erases any specific SSO buffer size.
+//   Because this is a base of the actual hash table, it is designed to be
+//   passed as a non-`const` reference or pointer.
+//
+// - A *table* type: the actual hashtable which derives from the base type and
+//   adds any desired SSO storage buffer. Beyond the physical storage, it also
+//   allows resetting the table to its initial state & allocated size, as well
+//   as copying and moving the table.
+//
+// For complete examples of the API design, see `set.h` for a hashtable-based
+// set data structure, and `map.h` for a hashtable-based map data structure.
+//
+// The hashtable design implemented here has several key invariants and design
+// elements that are essential to all three of the types above and the
+// functionality they provide.
+//
+// - The underlying hashtable use open addressing, a power-of-two table size,
+//   and quadratic probing rather than closed addressing and chaining.
+//
+// - The allocated table storage is organized into an array of one-byte metadata
+//   for every entry, followed by an array of storage for key/value pairs. The
+//   storage for an entry is an internal type that should not be exposed to
+//   users, and instead only the underlying keys and values.
+//
+// - The hash addressing and probing occurs over *groups* of entries rather than
+//   individual entries. These are not necessarily buckets of colliding entries,
+//   however, as they may be filled by probing with entries that have different
+//   initial hashed positions in the open addressed table.
```

Comment:
I think I see what you are trying to get at, but I think it is trying to answer a question the reader doesn't have (at least not at this point). And by trying to answer it, it creates more confusion (what do you mean this isn't for collisions?). It would be better to say how the groups are used in a positive way, and only after explaining how the probe sequence will look at additional groups if a group is full explain that as a consequence a group may have entries that were not the first choice for those entries. If you want, I can try and write some text to use.

Reply:
Here is the text I came up with: https://github.com/carbon-language/carbon-lang/pull/3940/files#r1614125444

Reply:
Yeah that's much better. Applied with some minor tweaks.